data/pigui/libs/modal-win.lua

```diff
@@ -0,0 +1,81 @@
+-- Copyright © 2008-2020 Pioneer Developers. See AUTHORS.txt for details
+-- Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+local ui = import 'pigui/pigui.lua'
+
+local modals = {}
+local modalStack = {}
+
+local ModalWindow = {}
+
+function ModalWindow.New(name, innerHandler, outerHandler, flags)
+	local self = {
+		name = name,
+		flags = flags or {"NoTitleBar", "NoResize", "AlwaysAutoResize", "NoMove"},
+		stackIdx = -1,
+		isOpen = false,
+		innerHandler = innerHandler,
+		outerHandler = outerHandler,
+	}
+
+	if not outerHandler then
+		self.outerHandler = function(drawPopupFn)
+			drawPopupFn()
+		end
+	end
+
+	setmetatable(self, {
+		__index = ModalWindow,
+		class = "UI.ModalWindow",
+	})
+
+	modals[name] = self
+
+	return self
+end
+
+function ModalWindow:open()
+	if self.stackIdx < 0 then
+		table.insert(modalStack, self)
+		self.stackIdx = #modalStack
+	end
+end
+
+function ModalWindow:close()
+	for i=#modalStack, self.stackIdx, -1 do
+		modalStack[i].stackIdx = -1
+		modalStack[i].isOpen = false
+		ui.closeCurrentPopup()
+		table.remove(modalStack, i)
+	end
+end
+
+local drawModals
+drawModals = function(idx)
```

Comment:
@vakhoir per the lua reference manual, `local function drawModals(idx)` is syntactic sugar for `local drawModals; drawModals = function(idx)`, so this construct is redundant.
Just FYI!

Reply:
Not exactly - it's not possible to do a recursive call if the function isn't already declared. The game crashes if you try to do it in a single line.

Reply:
That's weird, because the 5.2 manual explicitly states that here: https://www.lua.org/manual/5.2/manual.html#3.4.10

If it's crashing like that, there may be a deeper problem...

Reply:
Ah. I think I understand. Please prefer the use of `local function name()` rather than `local name = function()` - it's much easier to read and it gives more reliable results for Lua's debugger.